command/job_restart.go

```diff
@@ -0,0 +1,850 @@
+    between batches and waits for user input on how to proceed. If the answer
+    is a time duration all next batches will use this new value. Defaults to 0.
+
+  -fail-on-error
+    Fail command as soon as an allocation restart fails. By default errors are
+    accumulated and displayed when the command exits.
+
+  -group=<group-name>
+    Only restart allocations for the given group. Can be specified multiple
+    times. If no group is set all allocations for the job are restarted.
+
+  -no-shutdown-delay
+    Ignore the group and task 'shutdown_delay' configuration so there is no
+    delay between service deregistration and task shutdown or restart. Note
+    that using this flag will result in failed network connections to the
+    allocation being restarted.
+
+  -reschedule
+    If set, allocations are stopped and rescheduled instead of restarted
+    in-place. Since the group is not modified the restart does not create a new
+    deployment, and so values defined in 'update' blocks, such as
+    'max_parallel', are not taken into account. This option cannot be use with
+    '-task'.
+
+  -task=<task-name>
+    Specify the task to restart. Can be specified multiple times. If groups are
+    also specified the task must exist in at least of them. If no task is set
+    only tasks that are currently running are restarted. This option cannot be
+    used with '-all-tasks' or '-reschedule'.
+
+  -verbose
+    Display full information.
+`
+	return strings.TrimSpace(helpText)
+}
+
+func (c *JobRestartCommand) Synopsis() string {
+	return "Restart or reschedule allocations for a job"
+}
+
+func (c *JobRestartCommand) AutocompleteFlags() complete.Flags {
+	return mergeAutocompleteFlags(c.Meta.AutocompleteFlags(FlagSetClient),
+		complete.Flags{
+			"-all-tasks":         complete.PredictNothing,
+			"-batch-size":        complete.PredictAnything,
+			"-batch-wait":        complete.PredictAnything,
+			"-fail-on-error":     complete.PredictNothing,
+			"-no-shutdown-delay": complete.PredictNothing,
+			"-reschedule":        complete.PredictNothing,
+			"-task":              complete.PredictAnything,
+			"-verbose":           complete.PredictNothing,
+		})
+}
+
+func (c *JobRestartCommand) AutocompleteArgs() complete.Predictor {
+	return complete.PredictFunc(func(a complete.Args) []string {
+		client, err := c.Meta.Client()
+		if err != nil {
+			return nil
+		}
+
+		resp, _, err := client.Search().PrefixSearch(a.Last, contexts.Jobs, nil)
+		if err != nil {
+			return []string{}
+		}
+		return resp.Matches[contexts.Jobs]
+	})
+}
+
+func (c *JobRestartCommand) Name() string { return "job restart" }
+
+func (c *JobRestartCommand) Run(args []string) int {
+	c.ui = &cli.ConcurrentUi{Ui: c.Ui}
+
+	// Handle SIGINT to prevent accidental cancellations.
+	// activeCh is blocked while a signal is being handled.
+	activeCh := make(chan interface{})
+	sigsCh := make(chan os.Signal, 1)
+	signal.Notify(sigsCh, syscall.SIGINT)
```

Comment:
I think the problem in this case is that the user may choose [not to cancel the command](https://github.com/hashicorp/nomad/pull/16278/files#diff-331d4a10a3b907d64a5d41fda14da628e6b34efd7f7a0095e7ba52782279ff10R789-R792), but by then the context would have cancelled already and so we would need to recreate it, which is kind of awkward 😬

Reply:
Starting the signal handler at the top of the command but not polling on `activeCh` creates some weird artifacts where if the user hits `Ctrl-C` early (perhaps the initial queries are hanging) the command will continue to run and log until we hit the main restart loop:

```
$ nomad job restart example
^C
Are you sure you want to stop the restart process?
Allocations not restarted yet will not be restarted. [y/N] y

Canceling job restart process
==> 2023-03-01T10:04:25-05:00: Restarting 2 allocations
```

I see where you're going with the signal handler here in terms of making it "friendly", and it looks like your implementation handles double-Ctrl-C correctly (immediate exit). But if we're going to do this, we should push the signal handler down to right before we enter the main restart loop so that we're only checking for signals on the parts we're actually prepared to interrupt.

Reply:
Oh nice catch. Yeah, that signal handler is only useful for the long-running restart loop  so it makes sense to only start it when the loop starts 👍